### ISSUE TITLE
Provide listeners with FlowChartWithState

### ISSUE THREAD

**Issue by itssamuelrowe** [MrBlenny/react-flow-chart#21]
I want to listen when a node is selected by the user. For some reason, I was not able to import the functions declared in container/actions.ts. So I copied the source code into a local file and then imported it.

```

export const onNodeSizeChange: IOnNodeSizeChange = ({ nodeId, size }) => (chart: IChart) => {
  chart.nodes[nodeId] = {
    ...chart.nodes[nodeId],
    size,
  }
}
```

For some reason the `chart` is null. This brings the app to an halt. Which goes to say, I was not able to setup the `FlowChart` with my own state because of the above mentioned reasons.

I'm using the FlowChartWithState right now. It would be nice if the state and event callbacks were separated. The `FlowChartWithState` class does not expose any listener callbacks. Alternatively, I tried using the `onClick` callback on a custom `NodeInner` to listen for events. Looks like the event propagation is being blocked.

In short, I'm not able to listen to events using the `FlowChartWithState` class. Please provide an update for this issue as soon as possible (the flow chart is the primary component in my project). I'm willing to contribute the project if it helps.

**Comment by MrBlenny** [comment]
I assume you saw the [External State](https://github.com/MrBlenny/react-flow-chart/blob/master/stories/ExternalReactState.tsx) example? 

My guess is there you've got a different callback signature and that is causing chart to be null. You should be able to rewrite the callbacks and add your hooks as is.

**Comment by itssamuelrowe** [comment]
I just copied whatever was in 'actions.ts' from the source code. I didn't modify anything. :(

**Comment by itssamuelrowe** [comment]
```
import { v4 } from 'uuid'
import {
  IChart, IOnCanvasClick, IOnCanvasDrop, IOnDeleteKey, IOnDragCanvas, IOnDragNode, IOnLinkCancel,
  IOnLinkComplete, IOnLinkMouseEnter, IOnLinkMouseLeave, IOnLinkMove, IOnLinkStart, IOnNodeClick,
  IOnNodeSizeChange, IOnPortPositionChange,
} from '@mrblenny/react-flow-chart';

/**
 * This file contains actions for updating state after each of the required callbacks
 */

export const onDragNode: IOnDragNode = (event, data, id) => (chart: IChart) => {
  const nodechart = chart.nodes[id]

  if (nodechart) {
    chart.nodes[id] = {
      ...nodechart,
      position: {
        x: data.x,
        y: data.y,
      },
    }
  }

  return chart
}

export const onDragCanvas: IOnDragCanvas = (event, data) => (chart: IChart): IChart => {
  chart.offset.x = data.x
  chart.offset.y = data.y
  return chart
}

export const onLinkStart: IOnLinkStart = ({ linkId, fromNodeId, fromPortId }) => (chart: IChart): IChart => {
  chart.links[linkId] = {
    id: linkId,
    from: {
      nodeId: fromNodeId,
      portId: fromPortId,
    },
    to: {},
  }
  return chart
}

export const onLinkMove: IOnLinkMove = ({ linkId, toPosition }) => (chart: IChart): IChart => {
  const link = chart.links[linkId]
  link.to.position = toPosition
  chart.links[linkId] = { ...link }
  return chart
}

export const onLinkComplete: IOnLinkComplete = ({ linkId, fromNodeId, fromPortId, toNodeId, toPortId }) =>
  (chart: IChart): IChart => {
    if ([fromNodeId, fromPortId].join() !== [toNodeId, toPortId].join()) {
      chart.links[linkId].to = {
        nodeId: toNodeId,
        portId: toPortId,
      }
    }
    return chart
  }

export const onLinkCancel: IOnLinkCancel = ({ linkId }) => (chart: IChart) => {
  delete chart.links[linkId]
  return chart
}

export const onLinkMouseEnter: IOnLinkMouseEnter = ({ linkId }) => (chart: IChart) => {
  // Set the link to hover
  const link = chart.links[linkId]
  // Set the connected ports to hover
  if (link.to.nodeId && link.to.portId) {
    if (chart.hovered.type !== 'link' || chart.hovered.id !== linkId) {
      chart.hovered = {
        type: 'link',
        id: linkId,
      }
    }
  }
  return chart
}

export const onLinkMouseLeave: IOnLinkMouseLeave = ({ linkId }) => (chart: IChart) => {
  const link = chart.links[linkId]
  // Set the connected ports to hover
  if (link.to.nodeId && link.to.portId) {
    chart.hovered = {}
  }
  return chart
}

export const onLinkClick: IOnLinkMouseLeave = ({ linkId }) => (chart: IChart) => {
  if (chart.selected.id !== linkId || chart.selected.type !== 'link') {
    chart.selected = {
      type: 'link',
      id: linkId,
    }
  }
  return chart
}

export const onCanvasClick: IOnCanvasClick = () => (chart: IChart) => {
  if (chart.selected.id) {
    chart.selected = {}
  }
  return chart
}

export const onDeleteKey: IOnDeleteKey = () => (chart: IChart) => {
  if (chart.selected.type === 'node' && chart.selected.id) {
    const node = chart.nodes[chart.selected.id]
    // Delete the connected links
    Object.keys(chart.links).forEach((linkId) => {
      const link = chart.links[linkId]
      if (link.from.nodeId === node.id || link.to.nodeId === node.id) {
        delete chart.links[link.id]
      }
    })
    // Delete the node
    delete chart.nodes[chart.selected.id]
  } else if (chart.selected.type === 'link' && chart.selected.id) {
    delete chart.links[chart.selected.id]
  }
  if (chart.selected) {
    chart.selected = {}
  }
  return chart
}

export const onNodeClick: IOnNodeClick = ({ nodeId }) => (chart: IChart) => {
  if (chart.selected.id !== nodeId || chart.selected.type !== 'node') {
    chart.selected = {
      type: 'node',
      id: nodeId,
    }
  }
  return chart
}

export const onNodeSizeChange: IOnNodeSizeChange = ({ nodeId, size }) => (chart: IChart) => {
  chart.nodes[nodeId] = {
    ...chart.nodes[nodeId],
    size,
  }
}

export const onPortPositionChange: IOnPortPositionChange = (nodeToUpdate, port, position) =>
  (chart: IChart): IChart => {
    const node = chart.nodes[nodeToUpdate.id]
    node.ports[port.id].position = {
      x: position.x,
      y: position.y,
    }

    chart.nodes[nodeToUpdate.id] = { ...node }

    return chart
  }

export const onCanvasDrop: IOnCanvasDrop = ({ data, position }) => (chart: IChart): IChart => {
  const id = v4()
  chart.nodes[id] = {
    id,
    position,
    type: data.type,
    ports: data.ports,
    properties: data.properties,
  }
  return chart
}

```



The only thing I've changed was the import statement.

**Comment by itssamuelrowe** [comment]
```
 render() {
        const { classes } = this.props;
        const stateActions = mapValues(actions, (func: any) =>
            (...args: any) => this.setState(func(...args)));
        
      return (
        <div className={ classes.root }>
          <FlowChart
                chart={ chart }
                Components={{
                    NodeInner: this.renderInner,
                }}
                callbacks={ stateActions }
            />
        </div>
        )
    }
```

This is the render function I've written.

**Comment by itssamuelrowe** [comment]
![Untitled](https://user-images.githubusercontent.com/19667887/57827576-76636980-77c5-11e9-98f4-247f0606bd4e.png)

This is the error that I'm getting.

**Comment by itssamuelrowe** [comment]
![Untitled](https://user-images.githubusercontent.com/19667887/57827658-c17d7c80-77c5-11e9-9d19-63f2c4fbd61f.png)

This is the error I'm getting.

**Comment by MrBlenny** [comment]
What is the initial state of your chart? You'll probably get an error like that if it is initialised as an empty object (and not the IChart shape).

**Comment by itssamuelrowe** [comment]
```

const chart = {
  offset: {
    x: 0,
    y: 0,
  },
  nodes: {
    node1: {
      id: 'node1',
      type: 'aspect',

      properties: {
          name: 'Input',
          type: 'onecube.core',
          index: 0
      },

      position: {
        x: 300,
        y: 100,
      },
      ports: {
        port1: {
          id: 'port1',
          type: 'input',
          properties: {
            value: 'yes',
          },
        },
        port2: {
          id: 'port2',
          type: 'output',
          properties: {
            value: 'no',
          },
        }
      },
    },
    node2: {
      id: 'node2',
      type: 'directFlow',
      position: {
        x: 500,
        y: 300,
      },
      ports: {
        port1: {
          id: 'port1',
          type: 'input',
        },
        port2: {
          id: 'port2',
          type: 'output',
        },
      },
    },
    node3: {
      id: 'node3',
      type: 'aspect',
      position: {
        x: 300,
        y: 900,
      },
      properties: {
          name: 'Login',
          type: 'onecube.authentication',
          index: 1
      },
      ports: {
        port1: {
          id: 'port1',
          type: 'input',
        },
        port2: {
          id: 'port2',
          type: 'output',
        },
      },
    },
    node4: {
      id: 'node4',
      type: 'directFlow',
      position: {
        x: 500,
        y: 1100,
      },
      ports: {
        port1: {
          id: 'port1',
          type: 'input',
        },
        port2: {
          id: 'port2',
          type: 'output',
        },
      },
    },
    node5: {
      id: 'node5',
      type: 'aspect',
      position: {
        x: 300,
        y: 1700,
      },
      properties: {
          name: 'Output',
          type: 'onecube.core',
          index: 2
      },
      ports: {
        port1: {
          id: 'port1',
          type: 'input',
        },
        port2: {
          id: 'port2',
          type: 'output',
        },
      },
    },
  },
  links: {
    link1: {
      id: 'link1',
      from: {
        nodeId: 'node1',
        portId: 'port2',
      },
      to: {
        nodeId: 'node2',
        portId: 'port1',
      },
    },
    link2: {
      id: 'link2',
      from: {
        nodeId: 'node2',
        portId: 'port2',
      },
      to: {
        nodeId: 'node3',
        portId: 'port1',
      },
    },
    link3: {
      id: 'link3',
      from: {
        nodeId: 'node3',
        portId: 'port2',
      },
      to: {
        nodeId: 'node4',
        portId: 'port1',
      },
    },
    link4: {
      id: 'link4',
      from: {
        nodeId: 'node4',
        portId: 'port2',
      },
      to: {
        nodeId: 'node5',
        portId: 'port1',
      },
    },
  },
  selected: {},
  hovered: {},
};

```

**Comment by MrBlenny** [comment]
Is that object assigned to `this.state` or is it something like`this.state.chart`? Only the former will work for those callbacks. Log out chart in the callback and you should figure out the issue.

**Comment by itssamuelrowe** [comment]
Yes! Thank you so much! I actually didn't set the initial value of `this.state`! I am new to React and friends.
It's working now!

**Comment by itssamuelrowe** [comment]
```
<ReactJsonThrottled
          src={chart}
          enableClipboard={false}
          style={{ overflowY: 'scroll', width: '400px' }}
        />
```

One last question. What is the purpose of this element? As shown in [External React State example](https://github.com/MrBlenny/react-flow-chart/blob/master/stories/ExternalReactState.tsx).

**Comment by MrBlenny** [comment]
It renders the json tree for demo purposes.
https://mrblenny.github.io/react-flow-chart/index.html?path=/story/state--external-react-state
